File: 字符串/KMP 重复的子字符串.go
package 字符串

//s能否由他的某一子串重复n次构成
func repeatedSubstringPattern(s string) bool {
	if len(s) == 1 {
		return false
	}
	next := getNextG(s)
	x := next[len(next)-1] //x表示字符串的 最长相同前后缀长度
	//数组长度减去最长相同前后缀的长度相当于是第一个周期的长度，也就是一个周期的长度，
	//如果这个周期可以被整除，就说明整个数组就是这个周期的循环。
	y := len(next) - x //y表示一个周期的长度

	if x == 0 || len(next)%y != 0 {
		return false
	} else {
		return true
	}
}

func getNextG(s string) []int {
	next := make([]int, len(s))
	next[0] = 0
	j := 0 //j表示前缀末尾
	i := 1 //i表示后缀末尾
	for i = 1; i < len(s); i++ {
		for j > 0 && s[j] != s[i] {
			j = next[j-1]
		}
		if s[j] == s[i] {
			j++
		}
		next[i] = j
	}
	return next
}

File: 字符串/KMP重复的子字符串.java
package 字符串;


import java.util.Arrays;

public class KMP重复的子字符串 {
    public boolean repeatedSubstringPattern(String s) {
        var next = getNext(s);
        int n = s.length();
        if (n == 1) return false;
        //最长相等前缀 = 最长相等后缀
        //最小重复子串 = n - 最长相等前（后）缀
        var lenOfMinSub = n - next[n - 1];
        if (lenOfMinSub == n) return false;
        return n % lenOfMinSub == 0;
    }

    public static int[] getNext(String s) {
        int n = s.length();
        int[] next = new int[n];
        next[0] = 0;
        int j = 0;
        int i = 1;
        for (; i < n; i++) {
            while (j > 0 && s.charAt(j) != s.charAt(i)) {
                j = next[j - 1];
            }
            if (s.charAt(j) == s.charAt(i)) {
                j++;
            }
            next[i] = j;
        }
        return next;
    }

    public static void main(String[] args) {
        System.out.println(Arrays.toString(getNext("abac")));
    }

}
